### DISCUSSION TITLE
How to access a dependency that belongs to another package?

### DISCUSSION THREAD

**Discussion by cloudcompute** [pnpm/pnpm#4201] [Q&A]
The pnpm package manager is superb, very fast. yarn just sucks when it comes to managing a monorepo.

I have a monorepo with the following directory structure.

a. The workspace root directory has a package.json. The shared dev dependencies like eslint, typescript, prettier, eslint-plugins etc. are placed here.

b. Below root 2 sub directories: 'apps' and 'services'.

c. 'apps' has further two folders: 'api' and 'web' each having their own package.json

d. 'services' has a sub-folder named 'shared' with package.json It has functionality that is common to both api and web.

e. 'shared' has a dependency.. 'moment' library.

f. 'api' also needs an access to the 'moment' library..

g. As of now the the 'moment' dependency has been installed in the package.json of both 'api' and 'shared'.

But I want to keep such shared dependencies (moment in our case) at a single place i.e. in one package.json only. The main reason is if I need to upgrade them, I need to run pnpm upgrade from a single place only and not from multiple places.

Can I place the moment dependency in the root package.json and then access it any where in the workspace tree? If yes, what is the command? Is it okay to place dependencies (in addition to devDependencies) in the root?  I am asking this because yarn warns us to do so.

If not, a woraround could be place it in, let's say, 'shared' package.json. Can I create its link so that other packages can access it? 
What is the exact command to achieve this and from which directory it should be executed?

Thanks
Raman

**Comment by Jack-Works** [comment]
You can install it in the root package.json

**Reply by shivam-baghla** [reply]
@Jack-Works , no that's not the solution. If we need the package for only one particular app.

**Comment by cloudcompute** [comment]
When I tried adding the command, it gave me this warning

```
ERR_PNPM_ADDING_TO_ROOT  Running this command will add the dependency 
to the workspace root, which might not be what you want  - if you really meant it, 
make it explicit by running this command again with the -w flag
```

Looks like the pnpm does not want us to add dependencies in the root. Why so, may I ask?

After using the -w flag, the package has been installed in the root and the packages under the root can access it. But an eslint rule forbids the use of extraneous packages. 

All of these, yarn, pnpm, and eslint, prevent us to add a dependency in the root by default. Wondering.. why?

**Reply by djgerhab** [reply]
Disappointed to see no one answer this so far, I'm asking the same question.

**Reply by danshilm** [reply]
Seems to me, that in monorepos, adding a package to the root `package.json` should mean that _that_ package will be used throughout all the apps, and while that can be determined for sure in small monorepos, it definitely isn't, at the very least, a quick thing to be able to determine in larger monorepos, hence the warning.

I also like the warning; accidentally (and also when purposefully) trying to install a package in the root results in me having to stop and think about whether adding it to the root is indeed the correct choice.

**Reply by shivam-baghla** [reply]
@cloudcompute @djgerhab @danshilm , yes this occurs coz of mono-repo setup, you can change current dir to the particular app one to add dependency to that particular app in-stead of running command at root level.

**Comment by cloudcompute** [comment]
Installing in the root is not an answer. Can anyone help?

**Reply by hyzx86** [reply]
if you really meant it, make it explicit by running this command again with the -w flag (or --workspace-root)

**Reply by hyzx86** [reply]
`pnpm i  oidc-client -w`

**Reply by shivam-baghla** [reply]
@cloudcompute , I found a way to change the current directory to a particular app dir and then run the same command to add a dependency, and it worked!

**Reply by cassepipe** [reply]
@shivam-baghla  how ?

**Comment by gluxon** [comment]
I see this is an older discussion from 1.5 years ago, but noticing it now since it got a recent comment a few hours ago.

> g. As of now the the 'moment' dependency has been installed in the package.json of both 'api' and 'shared'.
>
> But I want to keep such shared dependencies (moment in our case) at a single place i.e. in one package.json only. The main reason is if I need to upgrade them, I need to run pnpm upgrade from a single place only and not from multiple places.

I'm hoping pnpm can accomplish this through [template catalogs](https://github.com/pnpm/rfcs/blob/714d1d81d86446412d3dc9a3fdea4144b4d080bb/text/0003-templates.md#catalog) in the future. 

The exact syntax is still in discussion, but I'm hoping we can do something like the following:

```json5
// catalog.json
{
  "name": "monorepo-catalog",
  "pnpm": {
    "template": "catalog"
  },
  "dependencies": {
    "moment": "^2.29.4"
  }
}
```

Other packages would be able to reference through:

```json5
// packages/web/package.json
{
  "name": "web",
  "pnpm": {
    "templates": {
      "catalog": {
        "default": "file:../../catalog.json"
      }
    }
  },
  "dependencies": {
    "moment": "catalog:default"
  }
}
```
